ksqldb-engine/src/test/java/io/confluent/ksql/serde/SerdeFeaturesFactoryTest.java

```diff
@@ -390,10 +398,101 @@ public void shouldNotConvertFormatForMulticolKeysWhenSanitizingIfDisallowed() {
         SerdeFeatures.of());
 
     // When:
-    final KeyFormat sanitized = SerdeFeaturesFactory.sanitizeKeyFormat(format, 2, false);
+    final KeyFormat sanitized = SerdeFeaturesFactory.sanitizeKeyFormat(format, MULTI_SQL_TYPES, false);
 
     // Then:
     assertThat(sanitized.getFormatInfo(), equalTo(FormatInfo.of(KafkaFormat.NAME)));
     assertThat(sanitized.getFeatures(), equalTo(SerdeFeatures.of()));
   }
+
+  @Test
+  public void shouldConvertKafkaFormatForSingleKeyWithPrimitiveType() {
+    // Given:
+    final KeyFormat format = KeyFormat.nonWindowed(
+        FormatInfo.of(KafkaFormat.NAME),
+        SerdeFeatures.of());
+
+    // When:
+    final KeyFormat sanitized = SerdeFeaturesFactory.sanitizeKeyFormat(format, ImmutableList.of(SqlTypes.struct().build()), true);
+
+    // Then:
+    assertThat(sanitized.getFormatInfo(), equalTo(FormatInfo.of(JsonFormat.NAME)));
+    assertThat(sanitized.getFeatures(), equalTo(SerdeFeatures.of(SerdeFeature.UNWRAP_SINGLES)));
+  }
+
+  @Test
+  public void shouldConvertKafkaFormatForSingleKeyWithUnsupportedPrimitiveType() {
+    // Given:
+    final KeyFormat format = KeyFormat.nonWindowed(
+        FormatInfo.of(KafkaFormat.NAME),
+        SerdeFeatures.of());
+
+    // When:
+    final KeyFormat sanitized = SerdeFeaturesFactory.sanitizeKeyFormat(format, ImmutableList.of(SqlPrimitiveType.of(SqlBaseType.BOOLEAN)), true);
+
+    // Then:
+    assertThat(sanitized.getFormatInfo(), equalTo(FormatInfo.of(JsonFormat.NAME)));
+    assertThat(sanitized.getFeatures(), equalTo(SerdeFeatures.of(SerdeFeature.UNWRAP_SINGLES)));
+  }
+
+  @Test
+  public void shouldConvertDelimitedFormatForSingleKeyWithNonPrimitiveType() {
```

Comment:
Just curious: why should we convert for single column delimited format? I thought for either non-primitive or primitive types, single column delimited format does not need to be converted to json.

Reply:
delimited format does not support compound type for either single or multi-key case IMHO.

Reply:
+1 to @abbccdda 's rationale -- the delimited format does not support STRUCT, MAP, or ARRAY.

Reply:
Yeah I know; my question was primarily that, for a single column delimited format (which is a primitive type), why do we still need to convert to a json format, than just keeping it as a primitive typed format?

Reply:
Are you asking about this specific unit test or something else? In this specific unit test, there is a single, non-primitive type key (struct type), which is why the key format must be converted from DELIMITED to JSON.

Reply:
Sorry I should clarify myself a bit, my original confusions should be separated to two: 1) why delimited single-column non-primitive key should be converted to Json, and that's now cleared since DELIMITED does not support non-primitive types, 2) why delimited single-column primitive key should be converted to Json, and it seems we never convert in this case actually and I was just confused by not seeing a `shouldNotConvertDelimitedFormatForSingleKeyWithPrimitiveType` case.

Reply:
@guozhangwang for 2) I will add this as a new test case